src/lib/layout/network-menu/NetworkMenuBody.tsx

```diff
@@ -0,0 +1,192 @@
+/* eslint-disable sonarjs/cognitive-complexity */
+import { Accordion, Divider, Flex } from "@chakra-ui/react";
+import { observer } from "mobx-react-lite";
+import type { KeyboardEvent as ReactKeyboardEvent } from "react";
+import { useCallback, useMemo, useState } from "react";
+
+import { CHAIN_CONFIGS } from "config/chain";
+import { useCelatoneApp, useMobile, useSelectChain } from "lib/app-provider";
+import InputWithIcon from "lib/components/InputWithIcon";
+import { EmptyState } from "lib/components/state";
+import { useNetworkStore } from "lib/providers/store";
+import type { Option } from "lib/types";
+
+import { NetworkAccordion } from "./NetworkAccordion";
+import { NetworkAccodionPinned } from "./NetworkAccordionPinned";
+
+interface NetworkMenuBodyProps {
+  onClose: () => void;
+}
+
+const filterChains = (
+  chainIds: string[],
+  keyword: string,
+  type: "mainnet" | "testnet"
+) =>
+  chainIds
+    .filter((chain) => CHAIN_CONFIGS[chain]?.networkType === type)
+    .filter(
+      (network) =>
+        !keyword ||
+        CHAIN_CONFIGS[network]?.prettyName
+          .toLowerCase()
+          .includes(keyword.toLowerCase()) ||
+        network.toLowerCase().includes(keyword.toLowerCase())
+    );
+
+const getNextCursor = (
+  key: string,
+  current: Option<number>,
+  lastIndex: number
+) => {
+  switch (key) {
+    case "ArrowUp":
+      if (current === undefined) return lastIndex;
+      return current <= 0 ? lastIndex : current - 1;
+    case "ArrowDown":
+      if (current === undefined) return 0;
+      return current >= lastIndex ? 0 : current + 1;
+    default:
+      return undefined;
+  }
+};
+
+export const NetworkMenuBody = observer(({ onClose }: NetworkMenuBodyProps) => {
+  const isMobile = useMobile();
+  const selectChain = useSelectChain();
+  const { availableChainIds } = useCelatoneApp();
+  const { getPinnedNetworks } = useNetworkStore();
+
+  const [cursor, setCursor] = useState<number>();
+  const [keyword, setKeyword] = useState("");
+
+  // Get chains info
+  const pinnedNetworks = getPinnedNetworks();
+  const filteredPinnedNetworks = useMemo(() => {
+    if (!keyword) return [...pinnedNetworks];
+    return pinnedNetworks.filter(
+      (network) =>
+        CHAIN_CONFIGS[network.chainId]?.prettyName
+          .toLowerCase()
+          .includes(keyword.toLowerCase()) ||
+        network.chainId.toLowerCase().includes(keyword.toLowerCase())
+    );
+  }, [pinnedNetworks, keyword]);
```

Comment:
Can we utilize `filterChains` here? I think it's very similar by just making type argument to optional.